pkg/controller/dnsrecord/actuator.go

```diff
@@ -144,8 +145,21 @@ func (a *actuator) Migrate(ctx context.Context, dns *extensionsv1alpha1.DNSRecor
 
 func (a *actuator) getDomainName(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, dnsClient alicloudclient.DNS) (string, error) {
 	switch {
-	case dns.Spec.Zone != nil && *dns.Spec.Zone != "":
-		return *dns.Spec.Zone, nil
+	case dns.Spec.Zone != nil && *dns.Spec.Zone != "" && (dns.Status.Zone == nil || *dns.Status.Zone == "" || !zoneMatchesDomainName(*dns.Spec.Zone, *dns.Status.Zone)):
+		if isDomainName(*dns.Spec.Zone) {
+			return *dns.Spec.Zone, nil
+		}
+		// The value specified in dns.Spec.Zone is not a domain name, so assume it's a domain id,
+		// and try to determine the domain name by getting the name of the domain with this id
+		domainName, err := dnsClient.GetDomainName(ctx, *dns.Spec.Zone)
+		if err != nil {
+			return "", &controllererror.RequeueAfterError{
+				Cause:        fmt.Errorf("could not get DNS domain name for domain id %s: %+v", *dns.Spec.Zone, err),
+				RequeueAfter: requeueAfterOnProviderError,
+			}
+		}
+		a.logger.Info("Got DNS domain name", "domainName", domainName, "dnsrecord", kutil.ObjectName(dns))
+		return domainName, nil
 	case dns.Status.Zone != nil && *dns.Status.Zone != "":
```

Comment:
How about when a dns zone is switched to another account?
If `dns.Spec.Zone` is not specified, it seems we always go to `default` case.

Reply:
The `default` case will query the domains (via `DescribeDomains`) and will try to determine the correct domain name based on the dnsrecord name (using the "longest matching" approach implemented in `dnsrecord.FindZoneForName`). There is no change here. What is the actual issue that you see? Perhaps I don't actually understand what you mean by "a dns zone is switched to another account".

Reply:
It is a corner case. Please just ignore it in this PR.
Here is example:
- domain `example.com` is in AliCloud *account-a*
- DNS record is successfully reconciled, with status maintained correctly.
- domain `example.com` is designated in AliCloud *account-b*

It seems that the extension can't handle this case.

Reply:
If the domain is no longer accessible from the account, this would result in an error, with or without this PR. Since this PR introduced caching, `GetDomainName` might return an outdated result (still containing the domain), and so the error won't be `could not find DNS domain name for name ...`, but a different error when trying to manipulate an actual record. But in essence, I don't think this PR changes what happens in this case in a way that we would need to worry about.

Was this what you were concerned about, or perhaps I am still missing your point?

Reply:
Never mind, it is a corner case and it has no relationship with PR. We could discuss it if some issue happens.

Reply:
OK, resolving this discussion then, let me know if you think it should be addressed in a different PR.